source/crc.c

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "checksums.h"
+
+#include "aws/checksums/crc.h"
+#include "aws/common/byte_buf.h"
+
+PyObject *aws_py_checksums_crc32(PyObject *self, PyObject *args) {
+    (void)self;
+    // (void)args;
+
+    Py_buffer input_buf;
+    PyObject *py_previousCrc32;
+
+    if (!PyArg_ParseTuple(args, "s*O", &input_buf, &py_previousCrc32)) {
+        return PyErr_AwsLastError();
+    }
+
+    uint32_t previousCrc32 = PyLong_AsUnsignedLong(py_previousCrc32);
+
+    if (previousCrc32 == (uint32_t)-1 && PyErr_Occurred()) {
+        return NULL;
```

Comment:
Should we return PyErr_AwsLastError here (like above) instead of NULL?

Reply:
No, this is correct. I still don't quite understand why. @graebm python will automatically add a useful error in the case of NULL. Not sure exactly what the difference is between the two situations

Reply:
So, functions from python C extension will raise the error for python. But the function from our lib will not. In that case, we will need to raise the Python Error from our aws last error, which is `PyErr_AwsLastError` for. EG: `PyArg_ParseTuple` is a function from python C extension, you can safely return NULL at line 18.

Reply:
oh that makes sense, I'll change 
>        return PyErr_AwsLastError();
to NULL then to remove the confusion.

Reply:
you don't literally want to `return NULL;` because we need to cleanup the buffer before returning, so we'll probably use a `goto`. But yes, in the end, you just want NULL to get returned

Reply:
why goto over calling a cleanup function?

Reply:
If you use a goto you can put all your cleanup code in one place.
If you don't use a goto, you need to copy/paste the cleanup code everywhere you do a return... and then in the future the function changes so that there's more cleanup and someone needs to find every return and copy/paste the extra code into all the different places cleanup happens ... and they miss something somewhere

cleanup in C is easier to manage with a goto